### ISSUE TITLE
Add optional support for floating point registers

### ISSUE THREAD

**Issue by schveiguy** [dlang-community/mecca#2]
I loved the presentation at dconf, and would love to get my hands dirty with mecca, seeing how it works as an underlying i/o system for a web service (perhaps I could combine diamond with this?)

But the fibers need to support floating point registers. It seems a niche usage for the FP registers to be unsupported, and would not work well in the general case.

Can you think of a way to make this work as a compile-time option? That is, an option to everything that deals with fibers to save and restore the FP registers? And then I would recommend making that the default, and using the non-default "don't save the FP registers" in your codebase.

**Comment by dnadlinger** [comment]
Hmm – isn't mecca's fibre implementation is Posix/x86_64 only anyway, where the floating point registers are caller-save in the SysV ABI anyway?

(`core.thread.Fibre`s don't save XMM registers either, for the same reason.)

**Comment by Shachar** [comment]
More importantly, how do you get compile time flags to a library compiled by dub? Is that even possible?

**Comment by Shachar** [comment]
Going over the ABI, we need to preserve mxcsr and the x87 CW registers, and that's it. I am a bit wary of just adding them unconditionally, however, as accessing the FPU's registers is notoriously slower than the integer ones.

**Comment by Shachar** [comment]
As for which registers need saving, I found this in the ABI docs:

> The  control  bits  of  the MXCSR register  are  callee-saved  (preserved  across calls), while the status bits are caller-saved (not preserved).  The x87 status word register is caller-saved, whereas the x87 control word is callee-saved.

**Comment by schveiguy** [comment]
> how do you get compile time flags to a library compiled by dub?

I was not expecting to have to change how the library is compiled. BTW, you can do that by using version options in the dub config -- it will push these down to all the libraries you depend on.

I was thinking more of compile-time parameters to the functions you call. That is, I was expecting a way to select in the calling function whether you save the registers or not. I'm not sure where this goes, maybe on the event loop function? Or would it be fiber specific?

e.g.:

```D
spawnFiber!(SaveFP.yes)(...)
```

**Comment by dnadlinger** [comment]
> Going over the ABI, we need to preserve mxcsr and the x87 CW registers, and that's it.

Indeed, yes – I always forget about these. (In theory its only the control bits out of `mxcsr` as well, but of course that doesn't help. And IIRC `{LD, ST}MXCSR` are at least around 10 ops on most uarchs.)

@schveiguy: Note that druntime doesn't save these either. That is, if your application really requires them to be kept, you wouldn't be able to use `core.thread.Fiber` either.

---

As a side note, the D spec rather ominously states states that

> If the floating-point rounding mode is changed within a function, it must be restored before the function exits. If this rule is violated (for example, by the use of inline asm), the rounding mode used for subsequent calculations is undefined.

Whether switching out to a different fiber counts as exiting the function is debatable, but so is the intent of that restriction in the first place.

**Comment by schveiguy** [comment]
> Note that druntime doesn't save these either.

My reason for creating the issue is that @Shachar brought it up at his talk as a potential hazard of using the library. I assumed that meant that druntime fibers don't have the limitation. What the limitation is specifically, I'll claim complete ignorance (both on what does and should happen). From a user perspective, what I was asking for is something that "just works" without any caveats. If that's already the case, then by all means close this issue.

Maybe the next step then is for Shachar to clarify when you have to be cautious when using the mecca fiber system, and we can determine whether anything needs addressing.

**Comment by Shachar** [comment]
I brought up the limitations of the mecca code. If the druntime code does the same thing, all it means is that its limitations are not documented.

I don't want to close the bug, as I think the limitations are real. I'll try to figure out a way to solve this that would not be too expensive.

**Comment by Shachar** [comment]
What I'm planning for this is the following:
A fiber can all `fiberUsesFP`. If it does, it is guaranteed that its FP registers are correctly saved.

The reactor will keep track of which fiber's FP registered are currently loaded in the CPU. It will only save those if switching to a different fiber that also requires FP.

This way:
* If you're not using FP, you're only paying for the conditional (which will be cached, and therefor cheap).
* If you're only using FP in one fiber, no extra overhead at all.
* If you're using FP in more than one fiber, you pay the context switch only between those fibers that do use it.
* It is exceedingly cheap to set this flag when you start using FP and reset it when you're done. Since, usually, no context switch will happen in the middle, this costs next to nothing.

**Comment by schveiguy** [comment]
It definitely sounds like it will be efficient, but horribly prone to errors :) I can see people using FP and not setting the flag since they *know* there won't be a context switch, and then in the future, someone inserts some code that might context switch, but forget to set the flag.

Can you also provide a mechanism to just set the flag on fiber start for those who don't want to deal with the hassle?

**Comment by Shachar** [comment]
You can certainly call it once at he beginning of the fiber, and never touch the flag again. It will be horribly inefficient.

A more efficient solution would be to treat this flag like you do a lock. Either set and put a scope(exit) that resets, or use a supplied RAII wrapper.

If you know you don't context switch, and you want to enforce it (i.e. - future proof your code), feel free to check out [critical section](https://weka-io.github.io/mecca/docs/mecca/reactor/Reactor.criticalSection.html). It provides a solution to precisely this problem. This is something you can do even today.

**Comment by schveiguy** [comment]
One thing that is going to be very useful if this enhancement is added is a guide on what this actually does and when to use it. It should include a demonstration of how the floating point registers can be messed up, and how to use this mechanism to avoid it. It may be something we want to add to `core.thread.Fiber` as well if it can cause problems.
